cli/tests/unit/http_test.ts

```diff
@@ -1782,6 +1782,120 @@ Deno.test("upgradeHttp", async () => {
+        await tlsConn.write(
+          new TextEncoder().encode(
+            "bla bla bla\nbla bla\nbla\n",
+          ),
+        );
+        tlsConn.close();
+      }, 500);
+    }
+
+    const abortController = new AbortController();
+    const signal = abortController.signal;
+    const certFile = "cli/tests/testdata/tls/localhost.crt";
+    const keyFile = "cli/tests/testdata/tls/localhost.key";
+
+    const server = serveTls((req) => {
+      const p = Deno.upgradeHttp(req);
+
+      (async () => {
+        const [conn, firstPacket] = await p;
+        const buf = new Uint8Array(1024);
+        const firstPacketText = new TextDecoder().decode(firstPacket);
+        assertEquals(firstPacketText, "bla bla bla\nbla bla\nbla\n");
+        const n = await conn.read(buf);
+        assert(n != null);
+        const secondPacketText = new TextDecoder().decode(buf.slice(0, n));
+        assertEquals(secondPacketText, "bla bla bla\nbla bla\nbla\n");
+        abortController.abort();
+        conn.close();
+      })();
+
+      return new Response(null, { status: 101 });
+    }, { port: 4501, signal, keyFile, certFile });
+
+    await Promise.all([server, client()]);
+  },
+);
+
+Deno.test("upgradeHttp unix", {
+  permissions: { read: true, write: true },
+  ignore: Deno.build.os === "windows",
+}, async () => {
+  const filePath = Deno.makeTempFileSync();
+  const promise = deferred();
+
+  async function client() {
+    const unixConn = await Deno.connect({ path: filePath, transport: "unix" });
+    await unixConn.write(
+      new TextEncoder().encode(
+        "CONNECT server.example.com:80 HTTP/1.1\r\n\r\nbla bla bla\nbla bla\nbla\n",
+      ),
+    );
+    setTimeout(async () => {
+      await unixConn.write(
+        new TextEncoder().encode(
+          "bla bla bla\nbla bla\nbla\n",
+        ),
+      );
+      unixConn.close();
+      promise.resolve();
+    }, 500);
+    await promise;
+  }
+
+  const server = (async () => {
+    const listener = Deno.listen({ path: filePath, transport: "unix" });
+    for await (const conn of listener) {
+      const httpConn = Deno.serveHttp(conn);
+      const maybeReq = await httpConn.nextRequest();
+      assert(maybeReq);
+      const { request, respondWith } = maybeReq;
+      const p = Deno.upgradeHttp(request);
+
+      const promise = (async () => {
+        const [conn, firstPacket] = await p;
+        const buf = new Uint8Array(1024);
+        const firstPacketText = new TextDecoder().decode(firstPacket);
+        assertEquals(firstPacketText, "bla bla bla\nbla bla\nbla\n");
+        const n = await conn.read(buf);
+        assert(n != null);
+        const secondPacketText = new TextDecoder().decode(buf.slice(0, n));
+        assertEquals(secondPacketText, "bla bla bla\nbla bla\nbla\n");
+        conn.close();
+      })();
+
+      const resp = new Response(null, { status: 101 });
+      await respondWith(resp);
+      await promise;
+      httpConn.close();
+      break;
```

Comment:
Seems weird to have a for await loop that breaks after one iteration.

There's a lot of sloppy loops in the tests that need to be unrolled and simplified - I expect the line-of-code could be halved.

Reply:
Opened https://github.com/denoland/deno/issues/13988